### ISSUE TITLE
pin_memory may change the type of instance returned by collate_fn.

### ISSUE THREAD

**Issue by kdexd** [pytorch/pytorch#32150]
## 🐛 Bug

Setting `pin_memory=True` to dataloader sometimes converts custom classes to native python types.

## To Reproduce

I have a class representing batched data, which I inherit from `dict` as follows:

```python
class TrainingBatch(dict):
    def __init__(self, instances: List[TrainingInstance]):
        # Implement logic equivalent to `collate_fn` here.
        ...

    def to(self, 8args, **kwargs):
        # Implement casting and device transfer logic here.
        ...
```

Hence `collate_fn` simply looks like this:
```python
def collate_fn(instances: List[TrainingInstance]) -> TrainingBatch:
    return TrainingBatch(instances)
```

### Main issue:

I set `pin_memory=True` to the dataloader, and here is what I get:
```python
>>> batch = next(iter(dataloader))
>>> type(batch)
<class 'dict'>
```

## Expected behaviour

```python
>>> type(batch)
<class 'TrainingBatch'>
```

## Bug:

The bug is here: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/pin_memory.py#L56
This `hasttr(data, "pin_memory")` should be the first condition to be checked, so custom classes can implement their own `pin_memory` method.

## Environment

 - PyTorch Version (e.g., 1.0): v1.3.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): None
 - Python version: 3.6.8

**Comment by SsnL** [comment]
It is more about https://github.com/pytorch/pytorch/blob/8d472bab6b29877dd316807aee7fb006700d8c8f/torch/utils/data/_utils/pin_memory.py#L51 should use `type(data)`, right?

**Comment by kdexd** [comment]
Correct, I ended up to pointing to a possible solution rather than the actual bug. I think either way is fine. I vote to have `hasattr` first because it removes the first condition (`torch.Tensor`) too.

**Comment by kdexd** [comment]
@SsnL your solution assumes that all keys in the `Mapping` are tensors. I may choose to have some boolean flags, or generally native Python types which I use within the model aside of tensor operations.

I can open a PR for the fix once we converge on what to implement here.
